File: src/test/java/io/lsdconsulting/lsd/distributed/captor/http/ResponseCaptorShould.java
package io.lsdconsulting.lsd.distributed.captor.http;

import feign.Request;
import feign.RequestTemplate;
import feign.Response;
import io.lsdconsulting.lsd.distributed.captor.http.derive.PathDeriver;
import io.lsdconsulting.lsd.distributed.captor.http.derive.SourceTargetDeriver;
import io.lsdconsulting.lsd.distributed.captor.repository.InterceptedDocumentRepository;
import io.lsdconsulting.lsd.distributed.captor.repository.model.InterceptedInteraction;
import io.lsdconsulting.lsd.distributed.captor.repository.model.InterceptedInteractionFactory;
import io.lsdconsulting.lsd.distributed.captor.trace.TraceIdRetriever;
import org.junit.jupiter.api.Test;

import java.util.Collection;
import java.util.List;
import java.util.Map;

import static feign.Request.HttpMethod.GET;
import static java.nio.charset.Charset.defaultCharset;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;
import static wiremock.org.apache.commons.lang3.RandomStringUtils.randomAlphanumeric;

class ResponseCaptorShould {

    private final InterceptedDocumentRepository interceptedDocumentRepository = mock(InterceptedDocumentRepository.class);
    private final SourceTargetDeriver sourceTargetDeriver = mock(SourceTargetDeriver.class);
    private final TraceIdRetriever traceIdRetriever = mock(TraceIdRetriever.class);

    private final PathDeriver pathDeriver = new PathDeriver();
    private final InterceptedInteractionFactory interceptedInteractionFactory = new InterceptedInteractionFactory("profile");

    private final ResponseCaptor underTest = new ResponseCaptor(interceptedDocumentRepository, interceptedInteractionFactory, sourceTargetDeriver, pathDeriver, traceIdRetriever);

    private final String url = randomAlphanumeric(20);
    private final String body = randomAlphanumeric(20);
    private final String traceId = randomAlphanumeric(20);
    private final String target = randomAlphanumeric(20);
    private final Map<String, Collection<String>> requestHeaders = Map.of("b3", List.of(traceId), "Target-Name", List.of(target));

    private final Response response = Response.builder()
            .request(Request.create(GET, url, requestHeaders, body.getBytes(), defaultCharset(), new RequestTemplate( )))
            .build();

    @Test
    public void takeTraceIdFromRequestHeaders() {
        given(traceIdRetriever.getTraceId(eq(requestHeaders))).willReturn(traceId);

        final InterceptedInteraction interceptedInteraction = underTest.captureResponseInteraction(response);

        assertThat(interceptedInteraction.getTraceId(), is(traceId));
    }

    @Test
    public void deriveTargetFromRequestHeaders() {
       given(sourceTargetDeriver.deriveTarget(eq(requestHeaders), eq(url))).willReturn(target);

        final InterceptedInteraction interceptedInteraction = underTest.captureResponseInteraction(response);

        assertThat(interceptedInteraction.getTarget(), is(target));
    }
}

File: spring-web/src/test/kotlin/io/lsdconsulting/lsd/distributed/interceptor/captor/ResponseCaptorShould.kt
package io.lsdconsulting.lsd.distributed.interceptor.captor

import io.lsdconsulting.lsd.distributed.interceptor.captor.http.SourceTargetDeriver
import io.lsdconsulting.lsd.distributed.interceptor.persistence.RepositoryService
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.apache.commons.lang3.RandomStringUtils.randomAlphanumeric
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.`is`
import org.junit.jupiter.api.Test
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpRequest
import org.springframework.http.HttpStatus
import org.springframework.http.client.ClientHttpResponse
import java.io.IOException

internal class ResponseCaptorShould {
    private val repositoryService = mockk<RepositoryService>(relaxed = true)
    private val sourceTargetDeriver = mockk<SourceTargetDeriver>()
    private val httpRequest = mockk<HttpRequest>()
    private val clientHttpResponse = mockk<ClientHttpResponse>()
    private val httpHeaderRetriever = mockk<HttpHeaderRetriever>()

    private val underTest = ResponseCaptor(repositoryService, sourceTargetDeriver, httpHeaderRetriever, "profile")

    private val traceId = randomAlphanumeric(20)
    private val target = randomAlphanumeric(20)
    private val serviceName = randomAlphanumeric(20)
    private val path = randomAlphanumeric(20)
    private val requestHeaders = mapOf<String, Collection<String>>("b3" to listOf(traceId), "Target-Name" to listOf(target))
    private val responseHeaders = mapOf<String, Collection<String>>()

    @Test
    @Throws(IOException::class)
    fun `handle empty response body from delete request`() {
        val httpHeaders = mockk<HttpHeaders>()
        every { httpHeaders.contentLength } returns 10
        every { httpRequest.headers } returns httpHeaders
        every { clientHttpResponse.headers } returns httpHeaders
        every { clientHttpResponse.statusCode } returns HttpStatus.NO_CONTENT
        every { clientHttpResponse.body } returns "body".byteInputStream()
        every { sourceTargetDeriver.deriveServiceName(requestHeaders) } returns serviceName
        every { httpHeaderRetriever.retrieve(httpRequest) } returns requestHeaders
        every { httpHeaderRetriever.retrieve(any<ClientHttpResponse>()) } returns responseHeaders

        val (_, body) = underTest.captureResponseInteraction(httpRequest, clientHttpResponse, target, path, traceId, 10L)

        assertThat(body, `is`("body"))
    }

    @Test
    @Throws(IOException::class)
    fun `enqueue intercepted interaction on spring response`() {
        val httpHeaders = mockk<HttpHeaders>()
        every { httpHeaders.contentLength } returns 10
        every { httpRequest.headers } returns httpHeaders
        every { clientHttpResponse.headers } returns httpHeaders
        every { clientHttpResponse.statusCode } returns HttpStatus.NO_CONTENT
        every { clientHttpResponse.body } returns "".byteInputStream()
        every { httpHeaderRetriever.retrieve(httpRequest) } returns requestHeaders
        every { sourceTargetDeriver.deriveServiceName(requestHeaders) } returns serviceName
        every { httpHeaderRetriever.retrieve(any<ClientHttpResponse>()) } returns responseHeaders

        val interceptedInteraction = underTest.captureResponseInteraction(httpRequest, clientHttpResponse, target, path, traceId, 10L)

        verify { repositoryService.enqueue(interceptedInteraction) }
    }
}
